Analysis/Tasks/PWGHF/HFTreeCreatorD0ToKPi.cxx

```diff
@@ -0,0 +1,263 @@
+                  hf_cand::PxProng1,
+                  hf_cand::PyProng1,
+                  hf_cand::PzProng1,
+                  hf_cand::ImpactParameter0,
+                  hf_cand::ImpactParameter1,
+                  hf_cand::ErrorImpactParameter0,
+                  hf_cand::ErrorImpactParameter1,
+                  full::NSigTPC_Pi_0,
+                  full::NSigTPC_Ka_0,
+                  full::NSigTOF_Pi_0,
+                  full::NSigTOF_Ka_0,
+                  full::NSigTPC_Pi_1,
+                  full::NSigTPC_Ka_1,
+                  full::NSigTOF_Pi_1,
+                  full::NSigTOF_Ka_1,
+                  full::CandidateSelFlag,
+                  full::M,
+                  full::ImpactParameterProduct,
+                  full::CosThetaStar,
+                  full::Pt,
+                  full::P,
+                  full::CPA,
+                  full::CPAXY,
+                  full::Ct,
+                  full::Eta,
+                  full::Phi,
+                  full::Y,
+                  full::E,
+                  full::MCflag);
+
+DECLARE_SOA_TABLE(HfCandProng2FullEvents, "AOD", "HFCANDP2FullE",
+                  collision::BCId,
+                  collision::NumContrib,
+                  collision::PosX,
+                  collision::PosY,
+                  collision::PosZ,
+                  full::IsEventReject,
+                  full::RunNumber);
+
+DECLARE_SOA_TABLE(HfCandProng2FullParticles, "AOD", "HFCANDP2FullP",
+                  collision::BCId,
+                  full::Pt,
+                  full::Eta,
+                  full::Phi,
+                  full::Y,
+                  full::MCflag);
+
+} // namespace o2::aod
+
+/// Writes the full information in an output TTree
+struct CandidateTreeWriter {
+  Produces<o2::aod::HfCandProng2Full> rowCandidateFull;
+  Produces<o2::aod::HfCandProng2FullEvents> rowCandidateFullEvents;
+  Produces<o2::aod::HfCandProng2FullParticles> rowCandidateFullParticles;
+  void init(InitContext const&)
+  {
+  }
+  void process(aod::Collisions const& collisions,
+               aod::McCollisions const& mccollisions,
+               soa::Join<aod::HfCandProng2, aod::HfCandProng2MCRec, aod::HFSelD0Candidate> const& candidates,
+               soa::Join<aod::McParticles, aod::HfCandProng2MCGen> const& particles,
+               aod::BigTracksPID const& tracks)
+  {
+
+    // Filling event properties
+    rowCandidateFullEvents.reserve(collisions.size());
+    for (auto& collision : collisions) {
+      rowCandidateFullEvents(
+        collision.bcId(),
+        collision.numContrib(),
+        collision.posX(),
+        collision.posY(),
+        collision.posZ(),
+        0,
+        1);
+    }
+
+    // Filling candidate properties
+    rowCandidateFull.reserve(candidates.size());
+    for (auto& candidate : candidates) {
+
+#define FILL_TREE(CandFlag, FunctionSelection, FunctionInvMass, FunctionCosThetaStar, FunctionCt, FunctionY, FunctionE) \
```

Comment:
I would recommend using a lambda or a dedicated function since macros are not checking types and thus are bug-prone.

Reply:
Thanks Anton! Done!